stablehlo/dialect/StablehloOps.cpp

```diff
@@ -5670,6 +5721,32 @@ ParseResult parseSameOperandsAndResultType(OpAsmParser& parser,
                                             *typesRef.back());
 }
 
+// The following implementation is for SameOperandsAndResultType with variadic
+// input.
+void printVariadicSameOperandsAndResultType(OpAsmPrinter& p, Operation* op,
+                                            OperandRange operands,
+                                            TypeRange opTypes, Type result) {
+  return printSameOperandsAndResultTypeImpl(p, op, opTypes, result);
+}
+
+ParseResult parseVariadicSameOperandsAndResultType(
+    OpAsmParser& parser,
+    SmallVectorImpl<OpAsmParser::UnresolvedOperand>& operands,
+    SmallVectorImpl<Type>& opTypes, Type& result) {
+  // Insert a type for each operand. Need to do this since passing the type of
+  // a variadic op gives no indication of how many operands were provided.
```

Comment:
Oof this is unfortunate. Let me take a closer look at this during the next pass over this PR. I'll submit my review for now because there's already plenty of comments to address.

Reply:
I've been staring at how this function creates a temporary vector of pointers and realized that reuse is probably overrated in this case. Then I rewrote this function without leveraging `parseSameOperandsAndResultTypeImpl`, and I think I like the result better:

```c++
ParseResult parseVariadicSameOperandsAndResultType(
    OpAsmParser& parser,
    SmallVectorImpl<OpAsmParser::UnresolvedOperand>& operands,
    SmallVectorImpl<Type>& opTypes, Type& result) {
  llvm::SMLoc loc = parser.getCurrentLocation();

  Type type;
  if (parser.parseType(type)) {
    return failure();
  }

  // Handle if function type, all operand types did not match result type.
  if (auto fnType = type.dyn_cast<FunctionType>()) {
    if (fnType.getResults().size() != 1) {
      return parser.emitError(loc, "expected single output");
    }
    llvm::append_range(opTypes, fnType.getInputs());
    result = fnType.getResults()[0];
    return success();
  }

  // Handle bare types. ` : type` indicating all input/output types match.
  opTypes.append(operands.size(), type);
  result = type;
  return success();
}
```

Yes, there is some duplication with `parseSameOperandsAndResultTypeImpl` (more precisely, the parseType and dyn_cast and fnType.getResults().size() != 1 parts), but I think this version of code is easier to understand. What's yout take on this?

Reply:
Hmm.. So the _real_ advantage of reuse in printer/parser code in my opinion is the parity:

```
let x = printSameOperandsAndResultTypeImpl(input)
let y = parseSameOperandsAndResultTypeImpl(x)
assert(y == input)
```

I feel like there is extra gravity to code duplication in printer/parser code. Duplicating this code adds the overhead of "if we change anything about the printer code, we need to update the parser code in two places." I'll keep thinking on how to improve this.